### PR TITLE
Add BZAccessControl

### PR DESCRIPTION
**Pull request by Top-Kraken**
Implemented BZAccessControl and add full test scripts for the smart contracts

### CODE REVIEW

**Review by hitsuji-haneta — COMMENTED** [review]
@Top-Kraken 
Thank you so much for creating the new PR!
Your idea looks interesting 👍👍👍👍

Let me confirm your idea :)
In my understanding,
- `BZAccessControl.sol` is what provided as a module
- `AdminContract.sol` is just an example of the composite side
Is it correct?

If so, what is the responsibility of `InheritContract.sol`?
For now, in Bunzz, users can not create a new contract inheriting modules.
I'm a bit confused.

File: BZAccessControl/contracts/AdminContract.sol [hunk 1]
```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./BZAccessControl.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+contract AdminContract is Context {
```
Comment (kazunetakeda25) [inline BZAccessControl/contracts/AdminContract.sol:8]: Please add NatSpec

Comment (kazunetakeda25) [reply BZAccessControl/contracts/AdminContract.sol:8]: Contract and functions

File: BZAccessControl/contracts/BZAccessControl.sol [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+import "@openzeppelin/contracts/utils/Context.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "./interfaces/IBZAccessControl.sol";
+import "./interfaces/IBunzz.sol";
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+
+contract BZAccessControl is Context, Ownable, IBZAccessControl, IBunzz, ERC165 {
```
Comment (kazunetakeda25) [inline BZAccessControl/contracts/BZAccessControl.sol:12]: Missing NatSpec

File: BZAccessControl/hardhat.config.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+require("hardhat-gas-reporter");
+
+/** @type import('hardhat/config').HardhatUserConfig */
+module.exports = {
+  solidity: {
+    version: "0.8.9",
+    settings: {
+      optimizer: {
+        enabled: true,
+        runs: 100,
```
Comment (kazunetakeda25) [inline BZAccessControl/hardhat.config.js:13]: For modules, lets use default runs and compile version will be always latest => 0.8.17

File: BZAccessControl/hardhat.config.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      },
+    },
+  },
+  gasReporter: {
+    enabled: true,
+    currency: 'CHF',
```
Comment (kazunetakeda25) [inline BZAccessControl/hardhat.config.js:19]: currency will be USD by default

File: BZAccessControl/package.json [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "build": "npx hardhat compile",
+    "test": "npx hardhat test",
+    "deploy": "npx hardhat run scripts/deploy.js"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (kazunetakeda25) [inline BZAccessControl/package.json:13]: Our main license will be MIT

File: BZAccessControl/hardhat.config.js (hunk 3 above)
Comment (kazunetakeda25) [reply BZAccessControl/hardhat.config.js:13]: default runs = 200

File: BZAccessControl/contracts/AdminContract.sol (hunk 1 above)
Comment (Top-Kraken) [reply BZAccessControl/contracts/AdminContract.sol]: This is Mockup contract for test only.

File: BZAccessControl/contracts/BZAccessControl.sol (hunk 2 above)
Comment (Top-Kraken) [reply BZAccessControl/contracts/BZAccessControl.sol]: Update the smart contract and added NatSpec

File: BZAccessControl/hardhat.config.js (hunk 3 above)
Comment (Top-Kraken) [reply BZAccessControl/hardhat.config.js]: Updated hardhat configuration

File: BZAccessControl/package.json (hunk 5 above)
Comment (Top-Kraken) [reply BZAccessControl/package.json]: Updated license and other things in package.json

**Comment by Top-Kraken** [comment]
Hi @hitsuji-haneta @kazunetakeda25 

Thanks for your feedbacks. I updated the smart contracts, test scripts and structures.
Please review again.

Thank you.

**Review by kazunetakeda25 — APPROVED** [review]

**Review by kazunetakeda25 — APPROVED** [review]

**Review by kazunetakeda25 — APPROVED** [review]

**Review by hitsuji-haneta — COMMENTED** [review]
@Top-Kraken 
Thank you! And sorry for the late reply🙏
The code looks perfect👍
But I'm just wondering if users can understand how to use it only by the document (README).

Do you have a good idea to explain how to use it?

**Comment by Top-Kraken** [comment]
> @Top-Kraken Thank you! And sorry for the late reply🙏 The code looks perfect👍 But I'm just wondering if users can understand how to use it only by the document (README).
> 
> Do you have a good idea to explain how to use it?

Hi @hitsuji-haneta 
Thanks for your feedback! I will try to update README again so that users can understand easily.

Thank you.

**Comment by Top-Kraken** [comment]
> 

Yes, AdminContractMock is contract for test only. When we upload the module, we will not add that mock contract.
cc @hitsuji-haneta @kazunetakeda25

**Comment by Top-Kraken** [comment]
Hi @hitsuji-haneta 

I updated the README of AccessControl. Could you check and give me the feedbacks?

Thank you.

**Review by hitsuji-haneta — APPROVED** [review]
@Top-Kraken 
The document became much better!
Thank you so much!

**Pull request merged**